lib/src/main/java/com/couchbase/lite/NativeLibrary.java

```diff
@@ -23,54 +23,82 @@
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.math.BigInteger;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.MessageDigest;
+import java.util.Arrays;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 
+/**
+ * For extracting and loading native libraries for couchbase-lite-java.
+ */
 final class NativeLibrary {
     private static final String[] LIBRARIES = { "LiteCore", "LiteCoreJNI" };
 
-    private static final String LIBS_RES_PATH = "/libs";
+    private static final String LIBS_RES_BASE_DIR = "/libs";
 
-    private static final String TMP_DIR_NAME = "com.couchbase.lite.java";
+    private static final String TARGET_BASE_DIR = "com.couchbase.lite.java/native";
 
     private static final AtomicBoolean LOADED = new AtomicBoolean(false);
 
+    /**
+     * Extracts and loads native libraries.
+     */
     static void load() {
         if (LOADED.getAndSet(true)) { return; }
-
-        for (String lib : LIBRARIES) { load(lib); }
-    }
-
-    private static void load(String libName) {
         try {
-            final File libFile = extractLibrary(libName);
-            System.load(libFile.getAbsolutePath());
+            final String[] libPaths = Arrays.stream(LIBRARIES).map(lib -> getResourcePath(lib)).toArray(String[]::new);
+            final String targetDir = getTargetDirectory(libPaths);
+            for (String path : libPaths) {
+                System.load(extract(path, targetDir).getAbsolutePath());
+            }
```

Comment:
... and then this a `forEach`...

Reply:
It cannot propagate the exception out of the closure so stick with the regular for-loop is cleaner.